relayer/relays/parachain/beefy-listener.go

```diff
@@ -82,125 +85,65 @@ func (li *BeefyListener) Start(ctx context.Context, eg *errgroup.Group) error {
 	eg.Go(func() error {
 		defer close(li.tasks)
 
-		// Subscribe NewMMRRoot event logs and fetch parachain message commitments
-		// since latest beefy block
-		beefyBlockNumber, _, err := li.fetchLatestBeefyBlock(ctx)
-		if err != nil {
-			return fmt.Errorf("fetch latest beefy block: %w", err)
-		}
-
-		err = li.doScan(ctx, beefyBlockNumber)
-		if err != nil {
-			return fmt.Errorf("scan for sync tasks bounded by BEEFY block %v: %w", beefyBlockNumber, err)
-		}
+		// Add some randomness here in case one relayer is down and other relayers won't compete for
+		// that failed message at same time.
+		ticker := time.NewTicker(time.Second*30 + time.Duration(rand.Intn(10))*time.Second)
+		for {
+			beefyBlockNumber, _, err := li.fetchLatestBeefyBlock(ctx)
+			if err != nil {
+				return fmt.Errorf("fetch latest beefy block: %w", err)
+			}
 
-		err = li.subscribeNewMMRRoots(ctx)
-		if err != nil {
-			if errors.Is(err, context.Canceled) {
+			err = li.doScan(ctx, beefyBlockNumber)
```

Comment:
Some cleanup here remove the `subscribeNewMMRRoots`.

The reason is that when switching to a multiple relayers context with task scheduled in a round-robin manner.  In case when a relayer is down, `subscribeNewMMRRoots` does not make sense for avoid all relayers competing for the fail message at the same time.

It seems a periodical scan with some randomness make more sense here, each round only one message is processed by a relayer, submit anyway when that message is timeout without anyone pick it up.

Reply:
not sure i understand this - surely all relayers should still subscribe to getting new mmr roots, even if they're not actively going to send a message proof for a given message?

Reply:
> all relayers competing for the same message at the same time in case when a relayer is down

Mainly for address the issue above. And I think subscribe is not required in any case, we do periodical scan for beacon also.

Reply:
I am open to this change, but I think we need more upfront design and analysis before writing more code. I'm worried that introducing polling & randomness could result in undefined behavior.

Specifically, I think the following would help:
1. In high-level pseudocode, a step-by-step analysis showing how a relayer would
  * Detect new messages
  * Makes sure there is MMR root which can be used to generate proofs
  * Backs off if it is not a designated relayer for a message
2. A theoretical simulation in which there are three different relayers, and two of them are offline.

Reply:
https://www.notion.so/snowfork/Decentralized-relayer-24511a7ec58b48f6911b9cf1307012af

Reply:
I think instead of randomness we should have primary and secondary slots.

Scenario:
3 Relayers with Ids 0, 1 and 2.

A relayer would calculate their slots as follows for a message nonce:
Primary slot: `nonce % numRelayers == id`
Secondary slot: `nonce % numRelyars - 1 == id`
Tertiary Slot: `nonce % numRelayers - 2 == id`
... etc, for as many slots as there are relayers

If a nonce lands on the primary slot for a relayer it relays immediately.
If a nonce lands on secondary slot it sleeps 3 minutes, checks if the nonce is relayed, and if not relays it.
If a nonce lands on the tertiary slot, it sleeps 6 minutes, checks if the nonce is relayed, and if not relays it.
... etc, for as many slots as there are relayers.

This is deterministic and every relayer covers every other relayers nonce at least once in a given timeframe.

Reply:
Cool!  Though there is still concern in my mind does that mean the sleep time will be linear correlation to the number of the relayers? So if we deployed 5 relayers, the worst case one relayer could wait for 15min,  and 10 relayers for 30 mins. Is that timeout too long?

Or is there any significant drawback for current impl(i.e. polling period with some randomness and waiting for a fixed timeout), IMHO it's not complicated and relayer also covers each other as tests demonstrated.

Reply:
@alistair-singh I think your idea is good, the concern above(e.g. waiting too long when there is a large number of relayers and most of them are down) is a rare case and may not be a real problem.

So I add another PR in https://github.com/Snowfork/snowbridge/pull/1266 follow the idea, please review & check if that matches  the design here.